### PR TITLE
[DRAFT] test/docs: VF allocator test suite & docs (depends on #355)

### PR DESCRIPTION
**Pull request by wangchen615**
#### What type of PR is this?

- [x] feature (test suite)
- [x] documentation

#### What this PR does:

This PR **depends on #355** (feat: in-segment block allocation for VF) and adds a
comprehensive test suite + documentation for the VF allocator implementation
introduced there.

**New Additions:**

1. **C++ Unit Test Suite** (`torch_spyre/csrc/test_vf_allocator.cpp`)
   - 5 focused unit tests validating core allocator data structures
   - Tests for FreeInterval ordering, BlockInfo creation, SegmentInfo initialization
   - Memory alignment calculation (128-byte boundaries)
   - Free interval merging algorithm (coalescing)
   - Standalone framework requiring no external dependencies
   - All 5 tests passing

2. **Python Integration Tests** (`tests/test_vf_allocator_standalone.py`)
   - 16 comprehensive tests validating VF allocator at PyTorch level
   - Test categories:
     - Initialization & device management (3 tests)
     - Memory management & fragmentation (8 tests)
     - Data types & operations (5 tests)
   - All 16 tests passing with FLEX_DEVICE=VF

3. **Comprehensive Documentation**
   - `docs/VFAllocatorUnitTests.md` - C++ test guide with detailed explanations
   - `docs/VFAllocatorPythonTests.md` - Python test guide with scenarios and patterns
   - Unified build/run instructions from project root
   - Key testing patterns and design rationale

**Key Features:**

- Both test suites run from project root directory
- Built artifacts output to `build/` directory for consistency
- Tests validate:
  - Data structure initialization and invariants
  - Memory alignment to 128-byte boundaries
  - Fragment coalescing and memory reuse
  - Real-world allocation patterns
  - Multi-segment scenarios (8GB per segment, 8 handlers)
  - Stress testing (50+ allocation cycles, mixed sizes)

#### Which issue(s) this PR is related to:

Depends on #355 (feat: in-segment block allocation for VF)
Complements #355 allocator implementation with comprehensive testing

#### Special notes for your reviewer:

- Tests are self-contained and standalone (C++ tests require no external dependencies)
- All tests passing and verified:
  - C++: `g++ -std=c++17 -DTEST_VF_ALLOCATOR torch_spyre/csrc/test_vf_allocator.cpp -o build/test_vf_allocator && ./build/test_vf_allocator`
  - Python: `FLEX_DEVICE=VF python -m pytest tests/test_vf_allocator_standalone.py -v`
- Documentation provides clear patterns for adding future tests
- Tests use production-scale values (8GB segments, not toy sizes)

#### Does this PR introduce a user-facing change?

No - this adds developer-facing test infrastructure and documentation. Tests are internal
validation tools and don't affect end-user functionality.

#### Additional notes:

This PR focuses on testing and documentation. It:
- Does not modify core allocator logic
- Provides strong validation baseline for #355 changes
- Establishes testing patterns for future VF allocator enhancements
- Includes both unit-level (C++) and integration-level (Python/PyTorch) testing

### CODE REVIEW

File: torch_spyre/csrc/test_vf_allocator.cpp [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+  free_intervals.insert(new_range);
+
+  // Should have one merged interval [0, 300]
+  ASSERT_EQ(free_intervals.size(), 1);
+  auto merged = *free_intervals.begin();
+  ASSERT_EQ(merged.start, 0);
+  ASSERT_EQ(merged.end, 300);
+}
+
```
Comment (JRosenkranz) [inline torch_spyre/csrc/test_vf_allocator.cpp]: I think we should also include tests specifically using the SpyreAllocator here? Initialization, allocation, etc.

**Review by JRosenkranz — CHANGES_REQUESTED** [review]
This is a great first pass for tests. One suggestion is I think we can keep the docs a bit higher level as this document will grow as we add more tests. The specific test explanations can be kept in the documentation within the test code. Another thing we may want to add is the ability to know which allocator is being used by adding some hooks into torch_spyre for this, as just checking the environment variable being used is not enough.

File: tests/test_vf_allocator_standalone.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+import torch  # noqa: E402
+
+
+class TestVFAllocatorStandalone(unittest.TestCase):
+    """Standalone test suite for VF allocator - can be run without pytest."""
+
+    def test_vf_mode_detection(self):
+        """Test that VF mode is correctly detected from environment variable."""
+        self.assertEqual(os.environ.get("FLEX_DEVICE"), "VF")
```
[the remark this answers is not in the document]
Comment (wangchen615) [reply tests/test_vf_allocator_standalone.py]: Added VF mode verification through 128-byte alignment check. The enhanced `test_vf_mode_detection` now verifies:

1. Environment variable is correctly set
2. Tensor allocation succeeds on spyre device
3. Allocation is aligned to 128 bytes (VF mode characteristic)

```python
def test_vf_mode_detection(self):
    """Test that VF mode is correctly detected from environment variable.

    Note: Currently the allocator mode is determined by the FLEX_DEVICE environment
    variable at allocator initialization time. The allocator's use_pf flag is set
    based on this variable. A future enhancement would be to expose an API like
    `torch_spyre.get_allocator_mode()` to verify the allocator mode at runtime.
    See: https://github.com/torch-spyre/torch-spyre/pull/448 for discussion.
    """
    self.assertEqual(os.environ.get("FLEX_DEVICE"), "VF")
    x = torch.empty(10, device="spyre", dtype=torch.float16)
    self.assertEqual(x.device.type, "spyre")
    
    # VF mode verification: allocations should be aligned to 128 bytes
    storage_size = x.untyped_storage().nbytes()
    self.assertEqual(storage_size % 128, 0, "VF mode requires 128-byte alignment")
```

**Note:** A more direct verification would require exposing the allocator's `use_pf` flag through a Python API like `torch_spyre.get_allocator_mode()`. This could be added as a future enhancement.

## Status
✅ Implemented in commit b21ac91

File: torch_spyre/csrc/test_vf_allocator.cpp [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+#define TEST(name)                                      \
+  void test_##name();                                   \
+  struct Register_##name {                              \
+    Register_##name() {                                 \
+      TestFramework::register_test(#name, test_##name); \
+    }                                                   \
+  } register_##name;                                    \
+  void test_##name()
+
+#define ASSERT_EQ(a, b)                                            \
```
[the remark this answers is not in the document]
Comment (wangchen615) [reply torch_spyre/csrc/test_vf_allocator.cpp]: Added support for Google Test via the `USE_GTEST` compile flag. The test file now conditionally uses either the lightweight custom framework or gtest:

### Standalone Mode (default)
Uses custom lightweight framework for quick compilation without dependencies:
```bash
g++ -std=c++17 -DTEST_VF_ALLOCATOR torch_spyre/csrc/test_vf_allocator.cpp \
    -o build/test_vf_allocator
```

### Google Test Mode
Uses gtest for CI integration:
```bash
g++ -std=c++17 -DTEST_VF_ALLOCATOR -DUSE_GTEST \
    torch_spyre/csrc/test_vf_allocator.cpp \
    -lgtest -lgtest_main -pthread \
    -o build/test_vf_allocator_gtest
```

The code structure:
```cpp
#ifdef USE_GTEST
#include <gtest/gtest.h>
// gtest provides TEST, ASSERT_EQ, ASSERT_TRUE, etc.
#else
// Lightweight custom framework
class TestFramework { ... };
#define TEST(name) ...
#define ASSERT_EQ(a, b) ...
#define ASSERT_TRUE(cond) ...
#endif
```

**Design Rationale:** The custom framework allows developers to run tests quickly without gtest dependency. For CI/CD pipelines, the gtest mode provides standard test reporting and integration.

## Status
✅ Implemented in commit b21ac91

File: torch_spyre/csrc/test_vf_allocator.cpp (hunk 1 above)
Comment (wangchen615) [reply torch_spyre/csrc/test_vf_allocator.cpp]: After consideration, we chose to test the SpyreAllocator through the **Python test suite** rather than adding C++ integration tests. Here's the reasoning:

### Why Python Tests for SpyreAllocator?

1. **Already comprehensive** - The Python tests in `test_vf_allocator_standalone.py` exercise the real SpyreAllocator through PyTorch's tensor allocation interface:
   - `test_basic_allocation` - Tests allocator initialization and allocation
   - `test_allocation_alignment` - Tests 128-byte alignment (parameterized by dtype)
   - `test_memory_reuse` - Tests deallocation and reuse
   - `test_oom_error_handling` - Tests error handling
   - `test_memory_limit_then_free` - Tests allocation near limits

2. **End-to-end validation** - Testing through PyTorch's interface validates the full integration stack, not just isolated allocator methods.

3. **Avoids test-only code in production files** - C++ integration tests would require:
   - Exposing private members (`use_pf`, `segments`, etc.)
   - Adding a test-only friend class or interface to `spyre_mem.cpp`
   - This adds maintenance burden and potential security concerns

### What the C++ Tests Cover

The C++ unit tests focus on **data structure correctness**:
- `FreeIntervalOrdering` - Interval ordering for `std::set`
- `BlockInfoCreation` - Block metadata initialization
- `SegmentInfoCreation` - Segment configuration (8GB, 8 handlers)
- `AlignmentCalculation` - 128-byte alignment algorithm
- `FreeIntervalMerging` - Coalescing algorithm

These can be validated independently without runtime dependencies.

### Running SpyreAllocator Tests

```bash
# Python tests (recommended - tests real allocator)
FLEX_DEVICE=VF python -m pytest tests/test_vf_allocator_standalone.py -v

# C++ unit tests (data structures only)
g++ -std=c++17 -DTEST_VF_ALLOCATOR torch_spyre/csrc/test_vf_allocator.cpp \
    -o build/test_vf_allocator && ./build/test_vf_allocator
```

## Status
✅ SpyreAllocator tested via Python test suite (18 tests)
✅ C++ tests focus on data structure validation (5 tests)
ℹ️ No placeholder C++ integration tests added (cleaner design)

## Open for Discussion

This was our initial design choice based on the rationale above. However, we're open to adding C++ integration tests if you feel they provide value that the Python tests don't cover. 

Potential approaches if C++ integration tests are desired:
1. Add a test-only getter method to `SpyreAllocator` (e.g., `isVFMode()`)
2. Use a friend class pattern for test access
3. Create a separate test allocator wrapper

Let us know your preference and we can adjust accordingly.

File: docs/VFAllocatorPythonTests.md [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+- Memory is properly reused after deallocation
+- The allocator handles edge cases and stress scenarios
+
+These tests complement the C++ unit tests by validating the full integration with PyTorch's tensor allocation system.
+
+### Purpose
+- Validate tensor allocation and device management
+- Test memory reuse and fragmentation handling
+- Verify alignment requirements are met
+- Test realistic allocation patterns with varying tensor sizes
```
[the remark this answers is not in the document]
Comment (wangchen615) [reply docs/VFAllocatorPythonTests.md]: Based on @dgrove-oss's feedback, tensor layout variations primarily affect the allocation sizes rather than the VF allocator logic itself. The current parameterized tests with different dtypes already cover size variations.

**Current Coverage:**
- `test_allocation_alignment` - Tests multiple sizes with various dtypes
- `test_mixed_size_allocations` - Tests allocations from 10 to 100K elements
- `test_large_tensor_allocation` - Tests ~160MB allocations

**Future Enhancement (if needed):**
Once the tensor layout copy work is complete, we could add:
```python
def test_tensor_layouts(self):
    layouts = [
        SpyreTensorLayout.Dense,
        SpyreTensorLayout.Sparse,
        SpyreTensorLayout.SparseMulti,
    ]
    for layout in layouts:
        # Test allocation with specific layout
        pass
```

## Status
ℹ️ Acknowledged - covered by existing size variation tests per @dgrove-oss feedback

File: docs/VFAllocatorPythonTests.md [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+
+## Test Descriptions
+
+### 1. **test_vf_mode_detection**
+**Purpose:** Verify that VF mode is correctly detected from environment variable.
+
+**Implementation:**
+
+```python
+def test_vf_mode_detection(self):
```
[the remark this answers is not in the document]
Comment (wangchen615) [reply docs/VFAllocatorPythonTests.md]: This is a good suggestion. Currently, we verify VF mode indirectly through:
1. Environment variable check (`FLEX_DEVICE=VF`)
2. 128-byte alignment verification (VF mode characteristic)

**Recommended Future Enhancement:**
Add a Python-accessible API to query the allocator mode:

```cpp
// In module.cpp, add:
m.def("get_allocator_mode", []() {
    return SpyreAllocator::instance().use_pf ? "PF" : "VF";
});
```

Then in Python tests:
```python
def test_vf_mode_allocator(self):
    """Verify VF allocator is being used."""
    mode = torch_spyre.get_allocator_mode()
    self.assertEqual(mode, "VF")
```

**Why Not Implemented Now:**
- Requires exposing allocator internals through pybind11
- The `use_pf` flag is currently private in `SpyreAllocator`
- Would need coordination with Andrea's allocator implementation

**Current Workaround:**
The `test_vf_mode_detection` test includes a note about this limitation and uses alignment verification as an indirect check.

## Status
ℹ️ Acknowledged - documented as future enhancement
📝 See comment in `test_vf_mode_detection` referencing this discussion

File: docs/VFAllocatorPythonTests.md [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+- `test_interleaved_allocations` - Sparse deallocation patterns
+- `test_free_interval_merging` - Adjacent free block merging
+- `test_mixed_size_allocations` - Variable-sized allocations
+- `test_allocation_after_many_deallocations` - Stress testing
+- `test_concurrent_allocations` - Multiple simultaneous allocations
+
+### 3. **Data Type & Operation Testing** (5 tests)
+- `test_different_dtypes` - Multiple data type support
+- `test_very_small_allocation` - Minimum size allocations
+- `test_large_tensor_allocation` - Large memory allocations
```
[the remark this answers is not in the document]
Comment (wangchen615) [reply docs/VFAllocatorPythonTests.md]: Done! Added `test_oom_error_handling` to verify proper error handling for out-of-memory scenarios:

```python
def test_oom_error_handling(self):
    """Test error handling when requesting more memory than available.

    This test verifies that the allocator properly raises an exception
    when memory request exceeds available segment size.
    """
    # VF allocator has 12GB segments. Try to allocate more than segment size.
    segment_size_bytes = 12 * 1024 * 1024 * 1024  # 12GB
    oversized_elements = (segment_size_bytes // 2) + 1  # float16 = 2 bytes

    with self.assertRaises(RuntimeError) as context:
        torch.empty(oversized_elements + (1024 * 1024 * 1024), device="spyre", dtype=torch.float16)

    self.assertTrue(
        "allocation" in str(context.exception).lower()
        or "memory" in str(context.exception).lower()
        or "segment" in str(context.exception).lower(),
        f"Expected OOM-related error message, got: {context.exception}",
    )
```

This test ensures:
- Allocator raises `RuntimeError` for oversized allocations
- Error message contains relevant keywords (allocation/memory/segment)
- No crash or undefined behavior

## Status
✅ Implemented in commit b21ac91

**Comment by wangchen615** [comment]
> This is a great first pass for tests. One suggestion is I think we can keep the docs a bit higher level as this document will grow as we add more tests. The specific test explanations can be kept in the documentation within the test code. Another thing we may want to add is the ability to know which allocator is being used by adding some hooks into torch_spyre for this, as just checking the environment variable being used is not enough.

Simplified both test documentation files to be higher-level, moving detailed test explanations to docstrings in the test files themselves.

**Commit:** [`749fd9f`](https://github.com/torch-spyre/torch-spyre/commit/749fd9f655459bc03c41798f7f21612607180e27) - feat: add get_allocator_mode() API and simplify test docs

### Changes Made

| File | Before | After | Reduction |
|------|--------|-------|-----------|
| `docs/VFAllocatorPythonTests.md` | 802 lines | 117 lines | 85% |
| `docs/VFAllocatorUnitTests.md` | 409 lines | 123 lines | 70% |

### What Was Removed
- Per-test detailed code snippets and implementation examples
- Step-by-step scenario walkthroughs for each test
- Verbose "What It Tests" and "Why It Matters" sections for each test

### What Was Kept
- Overview and purpose
- Test categories (high-level bullet points)
- Build/run instructions
- Key testing patterns
- Comparison table between C++ and Python tests
- References

### New Approach
Each doc now includes a note pointing to the test files for detailed explanations:

```markdown
For detailed explanations of individual tests, see the docstrings and comments in the test file itself.
```

**Comment by wangchen615** [comment]
Closing — this was an early POC for VF allocator tests and documentation. The design has since changed significantly with PR #1665 (merged), which introduces graph-free DMA with FlexAllocator and CompositeAddress. The allocator architecture this PR was testing and documenting no longer reflects the current implementation.
